File: rxbinding/src/main/java/com/jakewharton/rxbinding/widget/RxAdapter.java
package com.jakewharton.rxbinding.widget;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.widget.Adapter;
import rx.Observable;

import static com.jakewharton.rxbinding.internal.Preconditions.checkNotNull;

/**
 * Static factory methods for creating {@linkplain Observable observables} for {@link Adapter}.
 */
public final class RxAdapter {
  /**
   * Create an observable of data change events for {@code adapter}.
   * <p>
   * <em>Note:</em> A value will be emitted immediately on subscribe.
   */
  @CheckResult @NonNull
  public static <T extends Adapter> Observable<T> dataChanges(@NonNull T adapter) {
    checkNotNull(adapter, "adapter == null");
    return Observable.create(new AdapterDataChangeOnSubscribe<>(adapter));
  }

  private RxAdapter() {
    throw new AssertionError("No instances.");
  }
}

File: rxbinding-kotlin/src/main/kotlin/com/jakewharton/rxbinding2/widget/RxAdapter.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.jakewharton.rxbinding2.widget

import android.support.annotation.CheckResult
import android.widget.Adapter
import com.jakewharton.rxbinding2.InitialValueObservable
import kotlin.Suppress

/**
 * Create an observable of data change events for `adapter`.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
inline fun <T : Adapter> T.dataChanges(): InitialValueObservable<T> = RxAdapter.dataChanges(this)
